File: backend/app/services/matcher.py
"""Keyword matching logic.

Performs case-insensitive whole-word substring matching against post content.
"""

import re


class KeywordMatcher:
    def match(self, content: str, keywords: list[str]) -> list[str]:
        """Return the subset of keywords found in content (case-insensitive)."""
        matched = []
        content_lower = content.lower()
        for keyword in keywords:
            pattern = re.escape(keyword.lower())
            if re.search(pattern, content_lower):
                matched.append(keyword)
        return matched

File: backend/src/services/matcher.js
'use strict';

/**
 * Returns the subset of keywords whose phrase appears (case-insensitive)
 * anywhere in content.
 */
function matchKeywords(content, keywords) {
  const lower = content.toLowerCase();
  return keywords.filter((kw) => lower.includes(kw.toLowerCase()));
}

module.exports = { matchKeywords };
